### ISSUE TITLE
Strange behavior of the application regarding it's locale

### ISSUE THREAD

**Issue by xGrAnDx** [dotnet/runtime#98818]
I've encountered strange behavior of my .net 8 console application (OS: Ubuntu 20).
Whenever I read application's locale with Thread.CurrentThread.CurrentCulture I get correct cultureInfo ("ru_RU" in my case).

However, there is native dll function which returns message based on the current application locale (akin to MessageFormat). 
This function determines current locale by calling `setlocale` system function.

I came to the conclusion that, judging by native function return value, locale returned from the `setlocale` call differs from the one i'm getting with Thread.CurrentThread.CurrentCulture.

In order to proof that I used P/Invoke to call `setlocale` from my application with result being "C.UTF-8" locale, despite my shell having correct one:

```
LANG=ru_RU.UTF-8
LANGUAGE=
LC_CTYPE="ru_RU.UTF-8"  
LC_NUMERIC="ru_RU.UTF-8"  
LC_TIME="ru_RU.UTF-8"  
LC_COLLATE="ru_RU.UTF-8"  
LC_MONETARY="ru_RU.UTF-8"  
LC_MESSAGES="ru_RU.UTF-8"  
LC_PAPER="ru_RU.UTF-8"  
LC_NAME="ru_RU.UTF-8"  
LC_ADDRESS="ru_RU.UTF-8"  
LC_TELEPHONE="ru_RU.UTF-8"  
LC_MEASUREMENT="ru_RU.UTF-8"  
LC_IDENTIFICATION="ru_RU.UTF-8"  
LC_ALL=
```

When I set locale to "ru_RU.utf8" directly by calling `setlocale` from my app I can get correct return values from the native function.

Trying to call my application with environmental variable (e.g. LC_MESSAGES="ru_RU.UTF-8" MyApplicationName) leads to nothing since native dll still detects locale as "C.UTF-8".

This brought me to the couple of questions:
1) Is there specific behavior of the .net core runtime regarding locales that i'm not fond of?
2) Is there a workaround to setup Environmental variables in such way that `setlocale` function correctly return my "actual" locale instead of default "C.UTF-8" one?

**Comment by Clockwork-Muse** [comment]
> 1. Is there specific behavior of the .net core runtime regarding locales that i'm not fond of?

Without having looked at the code involved, it's likely that C# is only changing its own, internal to the dotnet runtime, locale value.  The C/C++ runtime is a completely separate thing.

> 2. Is there a workaround to setup Environmental variables in such way that setlocale function correctly return my "actual" locale instead of default "C.UTF-8" one?

Given this [SO question](https://stackoverflow.com/q/12170488/812837), it appears that C/C++ always requires you to explicitly set the current locale.  The answers listed there seem to indicate that supplying a specific parameter to the function will activate the "environment" locale, but you have to ask for it:
```cpp
setlocale(LC_ALL, "");
```

That aside, for a variety of testing/cross-platform reasons, I would suggest evaluating whether you can explicitly pass the locale from C# as an extra parameter to any culture-aware functions you call.

**Comment by tarekgh** [comment]
.NET fully depends on ICU library for globalization support. For default locale, we use the ICU API [uloc_getDefault](https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/uloc_8h.html#a4efa16db7351e62293f8ef0c37aac8d2). You can see the code here https://github.com/dotnet/runtime/blob/973ceee20c4807c780f72e24a1859a2c695ddf39/src/native/libs/System.Globalization.Native/pal_locale.c#L139. 

`uloc_getDefault` calls setlocale with LC_MESSAGES on Unix-like systems. The .NET behavior is clearly documented in https://learn.microsoft.com/en-us/dotnet/fundamentals/runtime-libraries/system-globalization-cultureinfo-currentculture#how-a-threads-culture-is-determined

**Comment by xGrAnDx** [comment]
So, if I understand this correctly:
When I'm using `CultureInfo.CurrentCulture` .NET tries to read `CultureInfo.DefaultThreadCurrentCulture` property. If its `null`, runtime then tries to call ICU uloc_getDefault (in the form of `setlocale(LC_MESSAGES, NULL)` call)?

Then i can't explain what is happening here:

In my test app i have next code:
```
        IntPtr locale = SetLocale(6 /*LC_MESSAGES*/, IntPtr.Zero);
        var localeStr = Marshal.PtrToStringAnsi(locale);
        
        Console.WriteLine("DefaultThreadCurrentCulture: " + CultureInfo.DefaultThreadCurrentCulture);
        Console.WriteLine("CurrentCulture: " + CultureInfo.CurrentCulture);
        Console.WriteLine("Thread.CurrentThread.CurrentCulture: " + Thread.CurrentThread.CurrentCulture);
        Console.WriteLine("setlocale ret value: " + localeStr);
```
Output:
```

DefaultThreadCurrentCulture:
CurrentCulture: ru-RU
Thread.CurrentThread.CurrentCulture: ru-RU
setlocale ret value: C

```
`DefaultThreadCurrentCulture` is empty, `setlocale` return value is `C` locale, while CurrentCulture (which should be taken from ICU lib setlocale call) is "ru_RU".

Is it possible that `setlocale` call from the ICU library and direct call from my app are different somehow?

So, basically my problem revolves around mismatch between .NET `CultureInfo` ("ru_RU") and actual `setlocale` return ("C.utf-8").

I need to change the locale in such a way the function call `setlocale(LC_MESSAGES, NULL)` from my application returns `ru_RU` instead of `C.utf-8`.

**Comment by tarekgh** [comment]
Look at [setlocale](https://www.man7.org/linux/man-pages/man3/setlocale.3.html) to understand how it works. Looks to me in your environment the locale was not set (i.e. set to empty or `C`). That is why you are getting the result you are seeing. I am not sure what exactly you want to achieve by calling `setLocale`. Here is some more info may help you understand the behavior: 

You may try to call `setlocale(LC_MESSAGES, "ru_RU")` to ensure setting the locale to the locale you desire first. I expect the subsequent calls `setlocale(LC_MESSAGES, NULL)` should return you `ru_RU`. 

.NET calls ICU `uloc_getDefault` which eventually calls the following code in ICU:
https://github.com/unicode-org/icu/blob/4633254f9ea434bb48a30fbcc98812eb561ff4eb/icu4c/source/common/putil.cpp#L1624

Basically to retrieve the default locale, it does the following operations in order :
- call `setlocale(LC_MESSAGES, nullptr);` try next step if the returned locale is empty or `C` or `POSIX`
- call `getenv("LC_ALL");` go to next step if get empty locale
- call `getenv("LC_MESSAGES")` go to next step if get empty locale
- call `getenv("LANG")`

This is why .NET returning you `ru-RU` most likely because your environment not having setLocale configuraed with `ru_RU` and one of the environment variables `LC_ALL`,  `LC_MESSAGES`, or `LANG` is set to  `ru_RU`
